src/lib/holocene/api-pagination.svelte

```diff
@@ -0,0 +1,164 @@
+<script lang="ts">
+  import Alert from '$lib/holocene/alert.svelte';
+  import FilterSelect from '$lib/holocene/select/filter-select.svelte';
+  import Icon from '$lib/holocene/icon/icon.svelte';
+  import SkeletonTable from '$lib/holocene/skeleton/table.svelte';
+
+  import { createPaginationStore } from '$lib/stores/api-pagination';
+  import { perPageOptions } from '$lib/stores/pagination';
+
+  type T = $$Generic;
+  type PaginatedRequest = (
+    size: number,
+    token: string,
+  ) => Promise<{ items: any[]; nextPageToken: string }>;
+
+  export let isUnauthorizedError: (error: any) => boolean = () => false;
+  export let onFetch: () => Promise<PaginatedRequest>;
+
+  let store = createPaginationStore();
+  let error: any;
+
+  $: nextIndex = $store.nextIndex;
+  $: pageSize = $store.pageSize;
+
+  function clearError() {
+    if (error) error = undefined;
+  }
+
+  async function onPageChange(nextIndex: number) {
+    clearError();
+    store.setUpdating();
+    try {
+      const fetchData: PaginatedRequest = await onFetch();
+      const response = await fetchData(pageSize, $store.indexTokens[nextIndex]);
+      const { items, nextPageToken } = response;
+      store.setNextPageToken(nextPageToken, items);
+    } catch (err: any) {
+      error = err;
+    }
+  }
+
+  async function onPageSizeChange(pageSize: number) {
+    clearError();
+    store.resetPageSize();
+    try {
+      const fetchData: PaginatedRequest = await onFetch();
+      const response = await fetchData(
+        pageSize,
+        $store.indexTokens[$store.index],
+      );
+      const { items, nextPageToken } = response;
+      store.setNextPageToken(nextPageToken, items);
+    } catch (err: any) {
+      error = err;
+    }
+  }
+
+  $: {
+    onPageChange(nextIndex);
+  }
+
+  $: {
+    onPageSizeChange(pageSize);
+  }
+
+  $: isEmpty = $store.items.length === 0 && !$store.loading;
+</script>
+
+{#if isUnauthorizedError(error) && $$slots.unauthorized}
```

Comment:
This doesn't seem super extensible 😞 Would welcome any thoughts on a better way to go about this.